### ISSUE TITLE
Output caching - support for donut caching

### ISSUE THREAD

**Issue by nfplee** [dotnet/aspnetcore#47268]
I’m exploring the new output caching added in ASP.NET Core 7. I reached out on [GitHub](https://github.com/dotnet/aspnetcore/issues/27387#issuecomment-1111171922) before it was released about whether this would support donut caching (where the hole is not cached, e.g. a cart widget for an ecommerce site). I was assured it would but I can't get it to work.

I have created my app, added the services + middleware and it works fine when I use the OutputCache attribute on an action. However my widget is a view component and I can’t get the OutputCache attribute to work on either the class or the method.

I had hoped I could add a different duration on the view component and it would be clever enough to know to cache the hole (the widget) separately.

Please could you confirm whether this is supported or whether there are any plans to add support for this in the future?

**Comment by mitchdenny** [comment]
@sebastienros - did we ever get a generalized doughnut caching solution in .NET 7 output caching?

**Comment by mitchdenny** [comment]
@nfplee just for clarity what are you using for your page rendering, Razor Pages, Razor views in MVC, Blazor?

**Comment by nfplee** [comment]
> @nfplee just for clarity what are you using for your page rendering, Razor Pages, Razor views in MVC, Blazor?

I'm using MVC.

**Comment by javiercn** [comment]
@nfplee I believe the outputcaching support is for endpoints in general.

For MVC there is the cache tag helper. If you are rendering the view component within an MVC action, you should be able to take (in theory) advantage of the new Output caching, if you are doing it from inside a razor view, that won't be the case, I believe.

**Comment by nfplee** [comment]
@javiercn thanks but even using the `CacheTagHelper` on my view component it doesn't allow it to be cached separately. For example, if I have an `OutputCacheAttribute` on my action with a duration of 10 seconds and I have an expires on my view component's `CacheTagHelper` of 5 seconds, it will still cache the hole (the view component) for 10 seconds.

**Comment by sebastienros** [comment]
My recommendation would be to use output caching with a much lower duration than the view components caches, not the other way around which makes the lower caches useless. The same way if you had database cache duration that were smaller than view component caches it wouldn't be useful.

As for donut caching, my recommendation is to render these silos of html by using ajax calls. For instance a Cart Widget, which is inherently authenticated data, would be served from a subsequent request with a custom cache strategy, that is user-aware, or no caching at all.

**Comment by nfplee** [comment]
Thanks @sebastienros for getting back to me.

The problem with your recommendation is I'd prefer it to work with server-side rendered HTML. I managed to come up with a solution for a PHP site I built a few years ago. Click [here](https://github.com/nfplee/flixon/blob/main/src/Flixon/Http/Middleware/ResponseCacheMiddleware.php) to take a look. It's quite simple in that it treats the cart as a child request and wraps it in a container (in my case a cache tag) if the parent request is cached. Then if the parent request is retrieved from the cache it replaces all the containers and re-executes the pipeline.
